File: backend/app/schemas/project.py
"""Internal API-ready schemas for canonical Projects and leadership roles."""

from datetime import date, datetime, timedelta, timezone
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class ProjectRoleAssignmentResponse(BaseModel):
    id: UUID
    role_type: Literal["PROJECT_LEADERSHIP"] = "PROJECT_LEADERSHIP"
    source_system: str
    source_person_id: str | None = None
    display_name: str
    native_role: str
    canonical_role: str
    email: str | None = None
    phone: str | None = None
    source_url: str | None = None
    source_document_id: str | None = None
    provenance: dict
    is_current: bool
    first_observed_at: datetime
    last_observed_at: datetime
    ended_at: datetime | None = None

    model_config = {"from_attributes": True}


class ProjectResponse(BaseModel):
    id: UUID
    source_system: str
    external_project_id: str
    name: str | None = None
    country: str | None = None
    region: str | None = None
    project_status: str | None = None
    approval_date: date | None = None
    closing_date: date | None = None
    borrower: str | None = None
    implementing_agencies: list[str] | None = None
    source_url: str | None = None
    enrichment_status: str
    enrichment_last_attempted_at: datetime | None = None
    last_enriched_at: datetime | None = None
    roles: list[ProjectRoleAssignmentResponse] = Field(
        default_factory=list,
        validation_alias="role_assignments",
    )

    model_config = {"from_attributes": True}

    @model_validator(mode="after")
    def expose_stale_success(self) -> "ProjectResponse":
        if self.enrichment_status != "successful" or self.last_enriched_at is None:
            return self
        enriched_at = self.last_enriched_at
        if enriched_at.tzinfo is None:
            enriched_at = enriched_at.replace(tzinfo=timezone.utc)
        if enriched_at < datetime.now(timezone.utc) - timedelta(days=7):
            self.enrichment_status = "stale"
        return self


class ProjectContextRoleResponse(BaseModel):
    """Presentation-safe Project leadership evidence for Tender Details."""

    role_type: Literal["PROJECT_LEADERSHIP"] = "PROJECT_LEADERSHIP"
    source_system: str
    display_name: str
    native_role: str
    canonical_role: str
    email: str | None = None
    phone: str | None = None
    source_url: str | None = None
    first_observed_at: datetime
    last_observed_at: datetime
    ended_at: datetime | None = None

    model_config = {"from_attributes": True}


class ProjectContextProjectResponse(BaseModel):
    """Source-neutral, whitelisted Project context exposed through a Tender."""

    id: UUID
    source_system: str
    external_project_id: str
    name: str | None = None
    country: str | None = None
    region: str | None = None
    status: str | None = Field(default=None, validation_alias="project_status")
    approval_date: date | None = None
    closing_date: date | None = None
    borrower: str | None = None
    implementing_agencies: list[str] | None = None
    source_url: str | None = None
    enrichment_status: str
    last_successful_enrichment_at: datetime | None = Field(
        default=None,
        validation_alias="last_enriched_at",
    )
    source_freshness: Literal[
        "fresh",
        "stale",
        "incomplete",
        "unavailable",
        "pending",
    ] = "pending"

    model_config = {"from_attributes": True}

    @model_validator(mode="after")
    def derive_source_freshness(self) -> "ProjectContextProjectResponse":
        status = self.enrichment_status
        if status == "successful":
            enriched_at = self.last_successful_enrichment_at
            if enriched_at is None:
                self.source_freshness = "incomplete"
                return self
            if enriched_at.tzinfo is None:
                enriched_at = enriched_at.replace(tzinfo=timezone.utc)
            if enriched_at < datetime.now(timezone.utc) - timedelta(days=7):
                self.enrichment_status = "stale"
                self.source_freshness = "stale"
            else:
                self.source_freshness = "fresh"
        elif status == "stale":
            self.source_freshness = "stale"
        elif status == "partial":
            self.source_freshness = "incomplete"
        elif status in {"source_unavailable", "failed"}:
            self.source_freshness = "unavailable"
        else:
            self.source_freshness = "pending"
        return self


class TenderProjectContextResponse(BaseModel):
    """Canonical Project and separately classified current/history roles."""

    project: ProjectContextProjectResponse
    current_roles: list[ProjectContextRoleResponse] = Field(default_factory=list)
    historical_roles: list[ProjectContextRoleResponse] = Field(default_factory=list)

File: frontend/types/project.ts
export type ProjectSourceFreshness =
    | 'fresh'
    | 'stale'
    | 'incomplete'
    | 'unavailable'
    | 'pending';

export type ProjectCanonicalRole =
    | 'TASK_TEAM_LEADER'
    | 'CO_TASK_TEAM_LEADER'
    | 'PROJECT_TASK_MANAGER'
    | 'OTHER_PROJECT_ROLE'
    | string;

export interface ProjectContextRole {
    role_type: 'PROJECT_LEADERSHIP';
    source_system: string;
    display_name: string;
    native_role: string;
    canonical_role: ProjectCanonicalRole;
    email: string | null;
    phone: string | null;
    source_url: string | null;
    first_observed_at: string;
    last_observed_at: string;
    ended_at: string | null;
}

export interface ProjectContextProject {
    id: string;
    source_system: string;
    external_project_id: string;
    name: string | null;
    country: string | null;
    region: string | null;
    status: string | null;
    approval_date: string | null;
    closing_date: string | null;
    borrower: string | null;
    implementing_agencies: string[] | null;
    source_url: string | null;
    enrichment_status: string;
    last_successful_enrichment_at: string | null;
    source_freshness: ProjectSourceFreshness;
}

export interface TenderProjectContext {
    project: ProjectContextProject;
    current_roles: ProjectContextRole[];
    historical_roles: ProjectContextRole[];
}

export interface ProjectMetadataRow {
    label: string;
    value: string;
}

export type ProjectContextFailureKind =
    | 'no_project'
    | 'authorization'
    | 'endpoint_failure';

export const classifyProjectContextFailure = (status?: number): ProjectContextFailureKind => {
    if (status === 404) return 'no_project';
    if (status === 401 || status === 403) return 'authorization';
    return 'endpoint_failure';
};

export const projectRoleLabel = (
    role: Pick<ProjectContextRole, 'canonical_role' | 'native_role' | 'source_system'>,
    sourceDisplayName = 'Official project source',
) => {
    // Defense in depth: this technical Projects API field never claims TTL semantics.
    if (role.native_role.trim().toLowerCase() === 'teamleadname') {
        return `${sourceDisplayName} project team`;
    }
    if (role.canonical_role === 'TASK_TEAM_LEADER') return 'Task Team Leader';
    if (role.canonical_role === 'CO_TASK_TEAM_LEADER') return 'Co-Task Team Leader';
    if (role.canonical_role === 'PROJECT_TASK_MANAGER') return 'Task Manager';
    if (role.canonical_role === 'OTHER_PROJECT_ROLE') {
        return `${sourceDisplayName} project team`;
    }
    return 'Project role';
};

export const projectFreshnessMessage = (freshness: ProjectSourceFreshness) => {
    if (freshness === 'stale') return 'Project information may be outdated.';
    if (freshness === 'incomplete') return 'Some project information is unavailable.';
    if (freshness === 'unavailable') return 'Official project data is currently unavailable.';
    if (freshness === 'pending') return 'Project details are being prepared.';
    return null;
};

export const formatProjectDate = (value?: string | null) => {
    if (!value) return null;
    const date = new Date(value);
    if (Number.isNaN(date.getTime())) return null;
    return date.toLocaleDateString('en-US', {
        month: 'short',
        day: 'numeric',
        year: 'numeric',
        timeZone: 'UTC',
    });
};

export const projectMetadataRows = (project: ProjectContextProject): ProjectMetadataRow[] => {
    const rows: ProjectMetadataRow[] = [];
    const geography = [project.country, project.region].filter(Boolean).join(' / ');
    const approvalDate = formatProjectDate(project.approval_date);
    const closingDate = formatProjectDate(project.closing_date);
    const agencies = (project.implementing_agencies ?? []).filter((agency) => agency.trim()).join(', ');

    if (geography) rows.push({ label: 'Country / Region', value: geography });
    if (project.status?.trim()) rows.push({ label: 'Project Status', value: project.status });
    if (approvalDate) rows.push({ label: 'Project Approval', value: approvalDate });
    if (closingDate) rows.push({ label: 'Project Closing', value: closingDate });
    if (project.borrower?.trim()) rows.push({ label: 'Borrower', value: project.borrower });
    if (agencies) rows.push({ label: 'Implementing Agency', value: agencies });
    return rows;
};
